src/front/glsl/lex_pp.rs

```diff
@@ -0,0 +1,242 @@
+use super::{parser::Token, token::TokenMetadata, types::parse_type};
+use crate::FastHashMap;
+use pp_rs::{
+    pp::Preprocessor,
+    token::{Punct, Token as PPToken, TokenValue},
+};
+use std::collections::VecDeque;
+
+//#[derive(Clone, Debug)]
+pub struct LexerPP<'a> {
+    pp: Preprocessor<'a>,
+    pp_buf: VecDeque<PPToken>,
+}
+
+impl<'a> LexerPP<'a> {
+    pub fn new(input: &'a str, defines: &'a FastHashMap<String, String>) -> Self {
+        let mut pp = Preprocessor::new(input);
+        for (define, value) in defines {
+            pp.add_define(define, value).unwrap(); //TODO: handle error
+        }
+        LexerPP {
+            pp,
+            pp_buf: Default::default(),
+        }
+    }
+
+    pub fn next(&mut self) -> Option<Token> {
+        let mut meta = TokenMetadata {
+            line: 0,
+            chars: 0..0,
+        };
+        let pp_token = match self.pp_buf.pop_front() {
+            Some(t) => t,
+            None => match self.pp.next()? {
+                Ok(t) => t,
+                Err((err, loc)) => {
+                    meta.line = loc.line as usize;
+                    meta.chars.start = loc.pos as usize;
+                    //TODO: proper location end
+                    meta.chars.end = loc.pos as usize + 1;
+                    return Some(Token::Unknown((meta, format!("{:?}", err))));
```

Comment:
any reason we can't just propagate the PP error up? I.e. instead of doing `format!("{:?}", err))` could we just do put `err` directly in there (as a `#[source]`)?

Reply:
Not quite sure I follow? Return type for iterator is `Option<Token>`, so wasn't sure how to propagate the error, except turning it into a string and wrapping in `Token::Unknown`

Reply:
I'm talking about `Token::Unknown(meta, glslpp::Error)` variant. Can we do it?

Reply:
ahh yeah that might work ... though would make parser directly dependent on glslpp 🤔 Is that what `#[source]` could help with?

Reply:
The `glsl-in` enables `glslpp` right now in this PR, is this not the goal? So we already depend on glslpp here.

Reply:
Yeah, was more thinking on source level to have glslpp entirely encapsulated in `lex.rs`, but that is probably over complicating things at this point. I'll try making it work with passing the error on.